### ISSUE TITLE
TESTERS REQUIRED: Unit Testing Configurations

### ISSUE THREAD

**Issue by Jeffail** [Jeffail/benthos#230]
The [Benthos Lab](https://lab.benthos.dev/), as well as [serverless sync responses](https://docs.benthos.dev/serverless/lambda/#running-a-combination) helped stabilize some tooling for performing discrete executions of a Benthos pipeline. 

It's now possible to build a pipeline, execute it against discrete input documents and capture the result programmatically. This opens up the possibility of constructing a unit test framework for Benthos configs. The implementation will be very simple so this ticket is for discussing the flow of such a framework.

My thoughts right now are that a new subcommand `test` should be added to the Benthos executable, it could maybe work similar to `go test` in that you could execute across an entire tree of tests with `benthos test ./...`.

This subcommand would prompt Benthos to search for Benthos config files to test and, according to the format we define here, finds sample input data and expected output data for each respective config and executes them. If the expected output does not match the real output then the test fails, a useful summary is printed and the subcommand exits with status 1.

## Identifying Tests

Ideally we want to point Benthos to a directory tree and have it automatically identify which files are Benthos configs to be tested. This is slightly problematic as we don't have a custom extension for Benthos config files, they could be `.yaml`, `.yml`, `.json`, etc. These extensions could be shared with any other config type.

We also need a file format that is able to describe our test conditions. This must include our test input and expected output, but might also need to include other information at a later stage.

## Running Tests in Parallel

Tests might take a while to execute, and so we should ideally run tests in parallel. This might be problematic for certain cases so we need to expose this either as a test-level parameter or a subcommand level parameter.

**Comment by Jeffail** [comment]
# Proposal 1

Benthos searches for files with a particular signature, perhaps `*.benthos_test.(yaml|yml|json)`, these are test definitions and should always be accompanied by a corresponding Benthos config, which is the target of the test.

For example, if we were to have a Benthos config named `foo.yaml`, we could define a test for that config named `foo.benthos_test.yaml`. The label `benthos_test` is pretty large so it would be nice to find a shorter variant. Also I'm not 100% sold on a dot separator, so maybe it could be `foo_benthos_test.yaml`.

The test definition contains any number of test cases, each of which will be an array of input messages (a batch) and an array of arrays of expected output messages (>0 batches):

``` yaml
test_cases_v1:
  - inputs:
      - '{"foo":["bar","baz"]}'
    outputs:
      -
        - "bar"
        - "baz"
```a

## Selecting Processors For Each Test

When `benthos test` executes the target config (`foo.yaml`) by default it will only execute the processors defined within `pipeline.processors`. Sometimes users may wish to test processors associated with a particular input or output in their config, or maybe test only a subset of the processors in their `pipeline.processors` list.

To solve this requirement we could add a field `target_processors` at the test case level which accepts a [JSON Pointer](https://tools.ietf.org/html/rfc6901) to the target processors. It's default value would be `/pipeline/processors`. Defining your own might look like this:


``` yaml
test_cases_v1:
  - target_processors: '/input/broker/inputs/0/processors'
    inputs:
      - '{"foo":["bar","baz"]}'
    outputs:
      -
        - "bar"
        - "baz"
```

## Strict Parsing of Test Definitions

When parsing your test definition Benthos will be strict, any field in your definition that is not recognised will cause an error. This will help reduce the chance of your tests silently succeeding due to being ignored.

**Comment by Jeffail** [comment]
Amended test definition format for proposal 1:

``` yaml
parallel: true
tests:
  - name: example test
    target_processors: '/input/broker/inputs/0/processors'
    input_batch:
      - content: '{"foo":["bar","baz"]}'
        metadata:
          example_key: example metadata value
    output_batches:
      -
        - content_equals: "bar"
          metadata_equals:
            key: example_key
            value: bar
        - content_equals: "baz"
```

I've added explicit fields for definition input content and other things such as metadata keys. I've also changed the way in which expected output is defined. Instead of simply writing our expected output messages we can define condition types, each with their own argument types.

If the count of batches, or messages of a batch, does not match what's listed in `output_batches` then the test will fail. However, if we only want to check the existence of a message but do not care about its contents or metadata we could have an empty entry in the list:

``` yaml
    output_batches:
      -
        - {} # Just check message 0 of batch 0 exists
        - content_equals: "baz" # Check the contents of message 1 batch 0 batches "baz"
```

**Comment by Jeffail** [comment]
I'm playing around with an implementation on this branch: https://github.com/Jeffail/benthos/tree/feature/config-tests

**Comment by Jeffail** [comment]
I've got an implementation that I'm happy with, I'm planning to merge into master today under the EXPERIMENTAL tag. I plan to immediately start implementing tests for our configs at Meltwater so I'll have a chance to test this stuff in anger, but it would be great to have users from other orgs try this out and give feedback over the next few weeks.

**Comment by Jeffail** [comment]
Closing for now, if you have comments/requests for changing this spec then feel free to reopen. If issues are found please raise a new ticket.
